Editor: Quick Actions Foundation

samwhale opened the pull request:
## Context

Adds the foundation for Quick Actions.

## Summary

Adding the foundation for Quick Actions.

## Relevant Technical Choices

- Hooked into the highlight api so that we could focus panes and panels across the app. Added benefit of using a familiar api for this future work
- The `TabView` component was holding redundant state for the selected tab. This tab state is already kept in the `LibraryProvider` and `InspectorProvider` so I removed the redundant state from the `TabView` component.
- Added some forward refs where we needed them

## To-do

- [x] add icon menu from #7475 
- [x] add karma tests
- [x] ~position the quick action menu in a good way 😕. LTR doesn't work, but other than that it's ready.~ added in #7510

## User-facing changes

![quick-actions](https://user-images.githubusercontent.com/22185279/118026170-c2e2d080-b31d-11eb-83ac-d4db1db3f3d4.gif)

## Testing Instructions

1. Turn on the `Enable Quick Action Menus` experiment
2. Go to editor
3. Click on the quick action buttons. They should focus specific panels and focus the background element.
    - Note: there is some lag while changing tabs. The animation plays instantly but since there is lag the tab may not flash on the first click. We should fix this in a future ticket.
4. Click the background if not focused, should see the same quick actions. Quick actions should work the same.
5. Add any type of element and focus this element. Quick action menu should hide (will show later once the actions are added for the specific element type).

### QA

<!--
Not all changes require manual QA.
-->

<!-- ignore-task-list-start -->
- [x] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

1. Turn on the `Enable Quick Action Menus` experiment
2. Go to editor
3. Click on the quick action buttons. They should focus specific panels and focus the background element.
    - Note: there is some lag while changing tabs. The animation plays instantly but since there is lag the tab may not flash on the first click. We should fix this in a future ticket.
4. Click the background if not focused, should see the same quick actions. Quick actions should work the same.
5. Add any type of element and focus this element. Quick action menu should hide (will show later once the actions are added for the specific element type).

### UAT

<!--
Sometimes the testing instructions for UAT can differ from the ones for QA.
-->

<!-- ignore-task-list-start -->
- [x] UAT should use the same steps as above.
<!-- ignore-task-list-end -->

## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.
Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #6151

samwhale committed:
Still figuring out some stuff

samwhale committed:
Add constants to new file

samwhale committed:
Refactor inspector and library tab view
- remove need initialTab
- remove redundant `tab` state. Now only single source of truth in the library and inspector providers
- remove copy of tabs

samwhale committed:
Make it work with the library

samwhale committed:
Add reset of actions for base state

samwhale committed:
Add tests, move files around

samwhale committed:
Add context menu to play with

samwhale commented on assets/src/edit-story/components/library/libraryPanes.js:
```diff
@@ -17,21 +17,67 @@
-        <Pane key={id} isActive={id === tab} aria-labelledby={getTabId(id)} />
-      )
-  );
+  const mediaHighlight = useFocusHighlight(states.MEDIA);
+  const textHighlight = useFocusHighlight(states.TEXT);
+
+  return tabs.map(({ id }) => {
+    const paneProps = {
+      key: id,
+      isActive: id === tab,
+      'aria-labelledby': getTabId(id),
+    };
+
+    switch (id) {
+      case MEDIA.id:
+        return (
+          <MediaPane
+            {...paneProps}
+            css={mediaHighlight?.showEffect && styles.FLASH}
+          />
+        );
+      case MEDIA3P.id:
+        return <Media3pPane {...paneProps} />;
+      case SHAPES.id:
+        return <ShapesPane {...paneProps} />;
+      case TEXT.id:
+        return (
+          <TextPane
+            {...paneProps}
+            css={textHighlight?.showEffect && styles.FLASH}
+          />
+        );
+      case ELEMS.id:
+        return <ElementsPane {...paneProps} />;
+      case PAGE_TEMPLATES.id:
+        return <PageTemplatesPane {...paneProps} />;
+      default:
+        return null;
+    }
+  });
```
The `css` prop has babel replace the element with a styled component. Babel couldn't figure out how to do this since the panes are decided at runtime and aren't visible at buildtime. This makes it so that babel can see which components it needs to replace when the app is being transpiled.

samwhale commented on assets/src/edit-story/components/tabview/index.js:
```diff
@@ -173,33 +173,31 @@ function TabView({
-  initialTab,
   ...rest
 }) {
-  const [tab, setTab] = useState(initialTab || tabs[0]?.id);
```
This was redundant state. The tabs (and current tab) are already being tracked in the library and inspector providers.

samwhale committed:
Remove unused stuff

samwhale committed:
Add quick action menu

maxyinger commented on assets/src/edit-story/app/highlights/quickActions/useQuickActions.js:
```diff
@@ -0,0 +1,125 @@
+    (selectedElements.length === 0 && backgroundElement) ||
+    selectedElements[0]?.isBackground
+  ) {
+    return [
```
should we memoize these at all?

samwhale replied:
Yeah that would probably be good. This will be rerendering the quick tips menu a lot 🤷

maxyinger commented on assets/src/edit-story/components/library/libraryProvider.js:
```diff
@@ -94,6 +61,19 @@ function LibraryProvider({ children }) {
+  }));
+
+  useEffect(() => {
+    if (highlightedTab && LIBRARY_TAB_IDS.has(highlightedTab)) {
```
do we need to check that `highlightedTab` exists since we're already checking if its in the set?
```
if (LIBRARY_TAB_IDS.has(highlightedTab)) {
```

samwhale replied:
Lmao this also found a sneaky bug. Thanks for the catch!

samwhale committed:
Add karma tests
* sort of not really fix styling
* TODO: fix styling for real

samwhale commented on assets/src/edit-story/components/canvas/karma/quickActions.karma.js:
```diff
@@ -0,0 +1,69 @@
+  });
+
+  describe('no element selected', () => {
+    it(`clicking the \`${ACTION_TEXT.CHANGE_BACKGROUND_COLOR}\` button should select the background and open the design panel`, async () => {
+      // click quick menu button
+      await fixture.events.click(
+        fixture.editor.canvas.quickActionMenu.changeBackgroundColorButton
+      );
+
+      expect(
+        fixture.editor.inspector.designPanel.pageBackground
+      ).not.toBeNull();
+    });
+
+    it(`clicking the \`${ACTION_TEXT.INSERT_BACKGROUND_MEDIA}\` button should select the background and open the media tab in the library`, async () => {
+      // change the library pane so media isn't visible
+      await fixture.events.click(fixture.editor.library.shapesTab);
+
+      // click quick menu button
+      await fixture.events.click(
+        fixture.editor.canvas.quickActionMenu.insertBackgroundMediaButton
+      );
+
+      expect(fixture.editor.library.media).not.toBeNull();
+    });
+
+    it(`clicking the \`${ACTION_TEXT.INSERT_TEXT}\` button should select the background and open the text tab in the library`, async () => {
+      // click quick menu button
+      await fixture.events.click(
+        fixture.editor.canvas.quickActionMenu.insertTextButton
+      );
+
+      expect(fixture.editor.library.text).not.toBeNull();
+    });
```
Could probably combine these 3 tests into 1. May not need to test each one individually

BrittanyIRL replied:
i like the separation of concerns

maxyinger reviewed (commented):
Looks great so far!

samwhale commented on assets/src/edit-story/components/canvas/layout.js:
```diff
@@ -233,6 +233,14 @@ const NavNextArea = styled(NavArea).attrs({
   area: 'f',
 })``;
 
+const QuickActionsArea = styled(PaddedPage)`
+  grid-area: p;
+  position: absolute;
+  right: calc(-24px + var(--page-padding-px));
+  padding-left: 0;
+  padding-right: 0;
+`;
```
Works for now. While this is being reviewed I'm going to update the grid so that this fits in the right spot.

samwhale replied:
Moved to another ticket: #7510

BrittanyIRL commented on assets/src/edit-story/app/highlights/quickActions/useQuickActions.js:
```diff
@@ -0,0 +1,131 @@
+    [handleFocusPanel]
+  );
+
+  // Hide menu if there are multiple elements selected
```
👍

BrittanyIRL commented on assets/src/edit-story/components/canvas/displayLayer.js:
```diff
@@ -24,14 +24,17 @@ import { __ } from '@web-stories-wp/i18n';
 /**
  * Internal dependencies
  */
+import { useFeature } from 'flagged';
```
nit: external

BrittanyIRL commented on assets/src/edit-story/components/library/libraryProvider.js:
```diff
@@ -94,6 +61,19 @@ function LibraryProvider({ children }) {
+  useEffect(() => {
+    if (highlightedTab && LIBRARY_TAB_IDS.has(highlightedTab)) {
+      setTab(highlightedTab);
+      trackEvent('quick_action_tab_change', {
```
nice

BrittanyIRL reviewed (commented):
This looks so beautiful! 

I'm going to do a little bit of testing here after lunch and then approve.

samwhale committed:
Fix comment and unit tests

samwhale committed:
PR Fixes
- fix import position
- don't need to check for truthiness since we're checking in the set already

samwhale committed:
Revert to working state
* fix bug where set had `undefined` in it

samwhale committed:
Fixes all around
- add fix for tab view: doesn't need redundant state
- add fix for quick tips: don't need to click twice

samwhale committed:
Fix tests, add new karma test

BrittanyIRL reviewed (approved):
So great! Thank you for all of the tests and easy to read structure. 

Since this is behind a flag, you can just go ahead and merge it once the CI says it's ok instead of moving to QA.
